### PR TITLE
feat(payment): INT-4021 Add strategy for iDeal APM

### PR DESCRIPTION
**Pull request by FranciscoOnt**
## What?
Create a payment strategy for ideal that sends the bic to the backend.

## Why?
This field is mandatory for Ideal payment to be processed

## Sibling PR
https://github.com/bigcommerce/checkout-js/pull/541

## Testing / Proof
Pending

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

**Review by cmedinilla — APPROVED** [review]
LGTM!

File: src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import CheckoutcomCustomPaymentStrategy from '../checkoutcom-custom-payment-strategy';
+
+const CHECKOUTCOM_IDEAL_PAYMENT_METHOD = 'ideal';
+export default class CheckoutcomiDealPaymentStrategy extends CheckoutcomCustomPaymentStrategy {
+
+    protected async _executeWithoutHostedForm(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const { payment, ...order } = payload;
+        const paymentData = payment?.paymentData;
+
+        if (!payment || !paymentData) {
```
Comment (capsula4) [inline src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts:15]: `if (!paymentData)` should suffice

File: src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        }
+    }
+
+    private _createFormattedPayload(methodId: string, paymentData: PaymentInstrument): WithCheckoutcomiDealInstrument {
+        const formattedPayload: WithCheckoutcomiDealInstrument = { bic: '' };
+        const bic = 'bic' in paymentData
+            ? paymentData.bic
+            : '';
+
+        if (CHECKOUTCOM_IDEAL_PAYMENT_METHOD === methodId && bic) {
```
Comment (capsula4) [inline src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts:40]: is it possible for you to simplify this method?

```js
if (CHECKOUTCOM_IDEAL_PAYMENT_METHOD === methodId && 'bic' in paymentData) {
  return { bic: paymentData.bic }
}

return { bic: '' } // maybe even return {};

File: src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.spec.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+        strategy = new CheckoutcomiDealPaymentStrategy(
+            store,
+            orderActionCreator,
+            paymentActionCreator,
+            formFactory
+        );
+    });
+
+    it('submits order without payment data', async () => {
```
Comment (capsula4) [inline src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.spec.ts:94]: `... when methodId in unsupported`

File: src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.spec.ts [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+            const payload = getOrderRequestBody();
+
+            await strategy.initialize(initializeOptions);
+            await strategy.execute(payload);
+
+            expect(store.dispatch)
+                .toHaveBeenCalledWith(loadOrderAction);
+        });
+
+        it('redirects to target url when additional action redirect is provided', async () => {
```
Comment (capsula4) [inline src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.spec.ts:312]: many of these test seem to be testing the parent class. Maybe it's not worth having them as a change in the parent class would trigger broken tests in a few different places.

File: src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts (hunk 1 above)
Comment (FranciscoOnt) [reply src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts:15]: it should, but the linter requires `payment` to be validated before using it.

**Comment by FranciscoOnt** [comment]
@capsula4 , I did a cleanup of the tests ♻️

File: src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts (hunk 1 above)
Comment (capsula4) [reply src/payment/strategies/checkoutcom-custom/checkoutcom-ideal/checkoutcom-ideal-payment-strategy.ts:15]: right I see, I can see you are using it further down

**Review by capsula4 — APPROVED** [review]

**Pull request merged**